pkg/controller/k8s-worker-pool/k8s_worker_pool.go

```diff
@@ -0,0 +1,92 @@
+package k8sworkerpool
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	envVarK8sClientsPoolSize  = "K8S_CLIENTS_POOL_SIZE"
+	defaultK8sClientsPoolSize = 10
+)
+
+// NewK8sWorkerPool returns a new k8s workers pool.
+func NewK8sWorkerPool(log logr.Logger) (*K8sWorkerPool, error) {
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get in cluster kubeconfig - %w", err)
+	}
+
+	ctx, cancelContext := context.WithCancel(context.Background())
+	k8sWorkerPoolSize := readEnvVars(log)
+
+	return &K8sWorkerPool{
+		ctx:           ctx,
+		log:           log,
+		cancelContext: cancelContext,
+		kubeConfig:    config,
+		jobsQueue:     make(chan *K8sJob, k8sWorkerPoolSize), // each worker can handle at most one job at a time.
+		poolSize:      k8sWorkerPoolSize,
+	}, nil
+}
+
+func readEnvVars(log logr.Logger) int {
+	k8sClientsPoolSize, found := os.LookupEnv(envVarK8sClientsPoolSize)
+	if !found {
+		log.Info(fmt.Sprintf("env variable %s not found, using default value %d", envVarK8sClientsPoolSize,
+			defaultK8sClientsPoolSize))
+
+		return defaultK8sClientsPoolSize
+	}
+
+	value, err := strconv.Atoi(k8sClientsPoolSize)
+	if err != nil || value < 1 {
+		log.Info(fmt.Sprintf("env var %s invalid value: %s, using default value %d", envVarK8sClientsPoolSize,
+			k8sClientsPoolSize, defaultK8sClientsPoolSize))
+
+		return defaultK8sClientsPoolSize
+	}
+
+	return value
+}
+
+// K8sWorkerPool pool that creates all k8s workers and the assigns k8s jobs to available workers.
+type K8sWorkerPool struct {
+	ctx           context.Context
+	log           logr.Logger
+	cancelContext context.CancelFunc
+	kubeConfig    *rest.Config
+	jobsQueue     chan *K8sJob
+	poolSize      int
+}
+
+// Start function starts the k8s workers pool.
+func (pool *K8sWorkerPool) Start() error {
+	for i := 1; i <= pool.poolSize; i++ {
+		worker, err := newK8sWorker(pool.log, i, pool.kubeConfig, pool.jobsQueue)
+		if err != nil {
```

Comment:
What happens if for example worker #5 failed to be initialized?
Are previously created workers will be stopped? 
As far as I see the returned error from here cause the upper Runnable to return error and doesn't call workerPool.Stop()

Reply:
@OlegSternbergIBM here is what happens:
bundle_syncer is the one who calls `k8sWorkerPool.Start`. if start returns error at any stage we will get to this line:
https://github.com/open-cluster-management/leaf-hub-spec-sync/blob/k8s-worker-pool/pkg/controller/bundles/bundles_syncer.go#L51

this will return an error from Start function of the bundle_syncer and in addition, will cancel the context (we have `defer cancelContext()` and since context is canceled all running workers will stop. 

does it answer the question?

Reply:
yes, thanks

Reply:
if you feel it's more readable we I can add the stop call in error case

Reply:
@OlegSternbergIBM added the following commit just to be on the same side:
https://github.com/open-cluster-management/leaf-hub-spec-sync/commit/3a0b43e46c731b3fb8feaaebb35d55fb3185127d